fetcher/fetcher_test.go

```diff
@@ -0,0 +1,258 @@
+}
+
+func newTestDB() *testDB {
+	return &testDB{
+		storage: make(map[string][]byte),
+	}
+}
+
+func newTestDBWithValue() *testDB {
+	db := testDB{storage: make(map[string][]byte)}
+	for i := 0; i < 100; i++ {
+		if i%2 == 0 {
+			db.storage[strconv.Itoa(i)] = []byte("value")
+		}
+	}
+	return &db
+}
+
+func (db *testDB) GetValue(_ context.Context, key []byte) (value []byte, err error) {
+	val, ok := db.storage[string(key)]
+	if !ok {
+		return nil, database.ErrNotFound
+	}
+	return val, nil
+}
+
+func TestFetchDifferentKeys(t *testing.T) {
+	var (
+		require = require.New(t)
+		numTxs  = 100
+		f       = New(newTestDB(), numTxs, 4)
+		ctx     = context.TODO()
+		wg      sync.WaitGroup
+
+		l     sync.Mutex
+		cache = make(map[string]interface{})
+	)
+	wg.Add(numTxs)
+
+	for i := 0; i < numTxs; i++ {
+		stateKeys := make(state.Keys, (i + 1))
+		for k := 0; k < i+1; k++ {
+			// Generate different read keys
+			stateKeys.Add(ids.GenerateTestID().String(), state.Read)
+		}
+		txID := ids.GenerateTestID()
+		// Since these are all different keys, we will
+		// fetch each key from disk
+		require.NoError(f.Fetch(ctx, txID, stateKeys))
+		go func(txID ids.ID) {
+			defer wg.Done()
+			reads, storage, err := f.Get(txID)
+			require.NoError(err)
+			l.Lock()
+			for k := range reads {
+				cache[k] = nil
+			}
+			for k := range storage {
+				cache[k] = nil
+			}
+			l.Unlock()
+		}(txID)
+	}
+	wg.Wait()
+	require.NoError(f.Wait())
+	require.Len(cache, 5050)
+}
+
+func TestFetchSameKeys(t *testing.T) {
+	var (
+		require = require.New(t)
+		numTxs  = 100
+		f       = New(newTestDB(), numTxs, 4)
+		ctx     = context.TODO()
+		wg      sync.WaitGroup
+
+		l     sync.Mutex
+		cache = make(map[string]interface{})
+	)
+	wg.Add(numTxs)
+
+	for i := 0; i < numTxs; i++ {
+		stateKeys := make(state.Keys, (i + 1))
+		for k := 0; k < i+1; k++ {
+			// Generate the same keys
+			stateKeys.Add(strconv.Itoa(k), state.Read)
+		}
+		txID := ids.GenerateTestID()
+		// We are fetching the same keys, so we should
+		// be getting subsequnt requests from cache
+		require.NoError(f.Fetch(ctx, txID, stateKeys))
+		go func(txID ids.ID) {
+			defer wg.Done()
+			reads, storage, err := f.Get(txID)
+			require.NoError(err)
+			l.Lock()
+			for k := range reads {
+				cache[k] = nil
+			}
+			for k := range storage {
+				cache[k] = nil
+			}
+			l.Unlock()
+		}(txID)
+	}
+	wg.Wait()
+	require.NoError(f.Wait())
+	require.Len(cache, 100)
+}
+
+func TestFetchSameKeysSlow(t *testing.T) {
+	var (
+		require = require.New(t)
+		numTxs  = 25
+		f       = New(newTestDB(), numTxs, 4)
+		ctx     = context.TODO()
+		wg      sync.WaitGroup
+
+		l     sync.Mutex
+		cache = make(map[string]interface{})
+	)
+	wg.Add(numTxs)
+	for i := 0; i < numTxs; i++ {
+		stateKeys := make(state.Keys, (i + 1))
+		for k := 0; k < i+1; k++ {
+			// Generate the same keys
+			stateKeys.Add(strconv.Itoa(k), state.Read)
+		}
+		txID := ids.GenerateTestID()
+		if i%2 == 0 {
+			time.Sleep(1 * time.Second)
```

Comment:
Usually, it is bad practice to sleep during a test. You should instead use a channel to replicate the behavior you want.